openmc/tally_derivative.py

```diff
@@ -0,0 +1,168 @@
+from __future__ import division
+
+import sys
+from numbers import Integral
+from xml.etree import ElementTree as ET
+
+from six import string_types
+
+import openmc.checkvalue as cv
+
+
+# "Static" variable for auto-generated TallyDerivative IDs
+AUTO_TALLY_DERIV_ID = 10000
+
+def reset_auto_tally_deriv_id():
+    global AUTO_TALLY_ID
+    AUTO_TALLY_DERIV_ID = 10000
+
+
+class TallyDerivative(object):
+    """A material perturbation derivative to apply to a tally.
+
+    Parameters
+    ----------
+    derivative_id : Integral, optional
+        Unique identifier for the tally derivative. If none is specified, an
+        identifier will automatically be assigned
+
+    Attributes
+    ----------
+    id : Integral
+        Unique identifier for the tally derivative
+    variable : str
+        Accepted values are 'density', 'nuclide_density', and 'temperature'
+    material : Integral
+        The perturubed material
+    nuclide : str
+        The perturbed nuclide. Only needed for 'nuclide_density' derivatives.
+        Ex: 'Xe-135'
+
+    """
+
+    def __init__(self, derivative_id=None):
+        # Initialize Tally class attributes
+        self.id = derivative_id
+        self._variable = None
+        self._material = None
+        self._nuclide = None
+
+    def __deepcopy__(self, memo):
+        existing = memo.get(id(self))
+
+        # If this is the first time we have tried to copy this object, create a
+        # copy
+        if existing is None:
+            clone = type(self).__new__(type(self))
+            clone.id = self.id
+            clone.variable = self.variable
+            clone.material = self.material
+            clone.nuclide = self.nuclide
+
+            memo[id(self)] = clone
+
+            return clone
+
+        # If this object has been copied before, return the first copy made
+        else:
+            return existing
+
+    def __eq__(self, other):
+        if not isinstance(other, TallyDerivative):
+            return False
+
+        return (self.variable == other.variable and
+                self.material == other.material and
+                self.nuclide == other.nuclide)
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        return hash(repr(self))
+
+    def __repr__(self):
+        string = 'Tally Derivative\n'
+        string += '{0: <16}{1}{2}\n'.format('\tID', '=\t', self.id)
```

Comment:
No need to number the placeholders since we require Python 2.7+. Same goes for all below as well